crates/ere-jolt/src/lib.rs

```diff
@@ -24,53 +28,73 @@ pub struct JOLT_TARGET;
 impl Compiler for JOLT_TARGET {
     type Error = JoltError;
 
-    type Program = Program;
+    // Note that we use `PathBuf` as `Program` because the only way to pass elf
+    // information to `jolt::host::Program` is by setting elf path, so we don't
+    // bother to read it into memory here.
+    type Program = PathBuf;
 
     fn compile(
         workspace_directory: &Path,
         guest_relative: &Path,
     ) -> Result<Self::Program, Self::Error> {
-        let manifest_path = workspace_directory
-            .join(guest_relative)
-            .to_path_buf()
-            .join("Cargo.toml");
-        let package_name = package_name_from_manifest(&manifest_path).unwrap();
-        let mut program = Program::new(&package_name);
-        program.set_std(true);
-        program.set_manifest_path(manifest_path);
-
-        // TODO: Note that if this fails, it will panic which is why it doesn't return a Result.
-        program.build(DEFAULT_TARGET_DIR);
-
-        Ok(program)
+        let guest_dir = workspace_directory.join(guest_relative);
+
+        // Change current directory for `Program::build` to build guest program.
+        set_current_dir(&guest_dir).map_err(|source| CompileError::SetCurrentDirFailed {
+            source,
+            path: guest_dir.to_path_buf(),
+        })?;
+
+        let package_name = package_name_from_manifest(Path::new("Cargo.toml"))?;
+
+        // Note that if this fails, it will panic, hence we need to catch it.
+        let elf_path = std::panic::catch_unwind(|| {
+            let mut program = Program::new(&package_name);
+            program.set_std(true);
+            program.build(DEFAULT_TARGET_DIR);
+            program.elf.unwrap()
+        })
+        .map_err(|_| CompileError::BuildFailed)?;
+
+        Ok(elf_path)
     }
 }
 
+#[derive(CanonicalSerialize, CanonicalDeserialize)]
+pub struct EreJoltProof {
+    proof: JoltHyperKZGProof,
+    public_outputs: Vec<u8>,
+}
+
 pub struct EreJolt {
-    program: <JOLT_TARGET as Compiler>::Program,
+    program: jolt::host::Program,
+    prover_preprocessing: JoltProverPreprocessing<4, jolt::F, jolt::PCS, jolt::ProofTranscript>,
+    verifier_preprocessing: JoltVerifierPreprocessing<4, jolt::F, jolt::PCS, jolt::ProofTranscript>,
+    _resource: ProverResourceType,
 }
 
 impl EreJolt {
-    pub fn new(
-        program: <JOLT_TARGET as Compiler>::Program,
-        _resource_type: ProverResourceType,
-    ) -> Self {
-        EreJolt { program }
+    pub fn new(elf_path: PathBuf, _resource: ProverResourceType) -> Self {
+        let mut program = Program::new("");
```

Comment:
When compiling in L52 it used the package name. Is it ok to use the empty string here?